hadoop-hdds/docs/content/design/decommissioning.md

```diff
@@ -0,0 +1,720 @@
+
+
+## Mainteneance mode in HDFS
+
+HDFS supports decommissioning and maintenance mode similar to Ozone. This is a quick description of the HDFS approach.
+
+The usage of HDFS maintenance mode:
+
+  * First, you set a minimum replica count on the cluster, which can be zero, but defaults to 1.
+  * Then you can set a number of nodes into maintenance, with an expiry time or have them remain in maintenance forever, until they are manually removed. Nodes are put into maintenance in much the same way as nodes are decommissioned.
+ * When a set of nodes go into maintenance, all blocks hosted on them are scanned and if the node going into maintenance would cause the number of replicas to fall below the minimum replica count, the relevant nodes go into a decommissioning like state while new replicas are made for the blocks.
+  * Once the node goes into maintenance, it can be stopped etc and HDFS will not be concerned about the under-replicated state of the blocks.
+  * When the expiry time passes, the node is put back to normal state (if it is online and heartbeating) or marked as dead, at which time new replicas will start to be made.
+
+This is very similar to decommissioning, and the code to track maintenance mode and ensure the blocks are replicated etc, is effectively the same code as with decommissioning. The one area that differs is probably in the replication monitor as it must understand that the node is expected to be offline.
+
+The ideal way to use maintenance mode, is when you know there are a set of nodes you can stop without having to do any replications. In HDFS, the rack awareness states that all blocks should be on two racks, so that means a rack can be put into maintenance safely.
+
+There is another feature in HDFS called "upgrade Domain" which allows each datanode to be assigned a group. By default there should be at least 3 groups (domains) and then each of the 3 replicas will be stored on different group, allowing one full group to be put into maintenance at once. That is not yet supported in CDH, but is something we are targeting for CDPD I believe.
+
+One other difference with maintenance mode and decommissioning, is that you must have some sort of monitor thread checking for when maintenance is scheduled to end. HDFS solves this by having a class called the DatanodeAdminManager, and it tracks all nodes transitioning state, the under-replicated block count on them etc.
+
+
+# Implementation
+
+
+## Datanode state machine
+
+`NodeStateManager` maintains the state of the connected datanodes. The possible states:
+
+  state             | description
+  ------------------|------------
+  HEALTHY           | The node is up and running.
+  STALE             | Some heartbeats were missing for an already missing nodes.
+  DEAD              | The stale node has not been recovered.
+  ENTER_MAINTENANCE | The in-progress state, scheduling is disabled but the node can't not been turned off due to in-progress replication.
```

Comment:
Should the in-progress state should be called ENTERING_MAINTENANCE to keep it similar to DECOMMISSIONING?

Reply:
Yes, that's a typo. Fixed, thanks.